From a0199697846ce314daf76541e0c94a5a568a33fb Mon Sep 17 00:00:00 2001
From: yangjunyi <yangjunyi5@huawei.com>
Date: Mon, 21 Nov 2022 20:06:21 +0800
Subject: [PATCH 0091/1398] [Huawei] drivers: hooks: add spinor driver vendor
 hooks

Offering: HULK
starry inclusion
category: bugfix
bugzilla: 187991

----------------------------------

Allow vendor modules to attach spinor driver hooks. This
patch introduce
vendor_spinor_erase_size/vendor_spi_nor_read_id hook.

Usage:

  static void vendor_spinor_foo1(
			void *data, struct spi_nor *nor, unsigned int *size)
  {
          pr_info("%s\n", __func__);
  }

  static void vendor_spinor_foo2(
			void *data, const struct flash_info **info)
  {
          pr_info("%s\n", __func__);
  }

  static int __init vendor_init(void)
  {
	int ret = 0;
	ret += register_trace_vendor_spinor_erase_size(&vendor_spinor_foo1, NULL);
	ret += register_trace_vendor_spi_nor_read_id(&vendor_spinor_foo2, NULL);

	return ret;
  }

  static void __exit vendor_exit(void)
  {
	unregister_trace_vendor_spinor_erase_size(&vendor_spinor_foo1, NULL);
	unregister_trace_vendor_spi_nor_read_id(&vendor_spinor_foo2, NULL);
  }

  module_init(vendor_init);
  module_exit(vendor_exit);

Signed-off-by: yangjunyi <yangjunyi5@huawei.com>
---
 drivers/hooks/Kconfig        | 10 ++++++++
 drivers/hooks/vendor_hooks.c |  8 +++++++
 drivers/mtd/spi-nor/core.c   |  9 +++++--
 include/trace/hooks/spinor.h | 46 ++++++++++++++++++++++++++++++++++++
 4 files changed, 71 insertions(+), 2 deletions(-)
 create mode 100644 include/trace/hooks/spinor.h

diff --git a/drivers/hooks/Kconfig b/drivers/hooks/Kconfig
index 9eaebaa17a97..2f21c8032c06 100644
--- a/drivers/hooks/Kconfig
+++ b/drivers/hooks/Kconfig
@@ -30,4 +30,14 @@ config VENDOR_LIBATA_HOOKS
         Allow vendor modules to attach libata driver hooks defined via
         DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
 
+config VENDOR_SPINOR_HOOKS
+          bool "Spinor driver Vendor Hooks"
+          depends on VENDOR_HOOKS && MTD_SPI_NOR
+          default n
+          help
+            Enable spinor driver vendor hooks
+
+            Allow vendor modules to attach spinor driver hooks defined via
+            DECLARE_HOOK or DECLARE_RESTRICTED_HOOK.
+
 endmenu
diff --git a/drivers/hooks/vendor_hooks.c b/drivers/hooks/vendor_hooks.c
index 102ca5397935..7899bb1487e8 100644
--- a/drivers/hooks/vendor_hooks.c
+++ b/drivers/hooks/vendor_hooks.c
@@ -10,6 +10,7 @@
 #include <trace/hooks/vendor_hooks.h>
 #include <trace/hooks/bonding.h>
 #include <trace/hooks/libata.h>
+#include <trace/hooks/spinor.h>
 
 /*
  * Export tracepoints that act as a bare tracehook (ie: have no trace event
@@ -23,3 +24,10 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_bond_check_dev_link);
 /* power on/off sata */
 EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_sata_reset);
 #endif
+
+#ifdef CONFIG_VENDOR_SPINOR_HOOKS
+/* change erase size by wanted_size */
+EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_spinor_erase_size);
+/* return default info table to do sfdp */
+EXPORT_TRACEPOINT_SYMBOL_GPL(vendor_spi_nor_read_id);
+#endif
diff --git a/drivers/mtd/spi-nor/core.c b/drivers/mtd/spi-nor/core.c
index e8146a47da12..d1f8baed30c0 100644
--- a/drivers/mtd/spi-nor/core.c
+++ b/drivers/mtd/spi-nor/core.c
@@ -21,7 +21,7 @@
 #include <linux/sched/task_stack.h>
 #include <linux/spi/flash.h>
 #include <linux/mtd/spi-nor.h>
-
+#include <trace/hooks/spinor.h>
 #include "core.h"
 
 /* Define max times to check status register before we give up. */
@@ -2085,7 +2085,11 @@ static const struct flash_info *spi_nor_read_id(struct spi_nor *nor)
 			return info;
 		}
 	}
-
+#ifdef CONFIG_VENDOR_SPINOR_HOOKS
+	trace_vendor_spi_nor_read_id(&info);
+	if (info != NULL)
+		return info;
+#endif
 	dev_err(nor->dev, "unrecognized JEDEC id bytes: %*ph\n",
 		SPI_NOR_MAX_ID_LEN, id);
 	return ERR_PTR(-ENODEV);
@@ -2586,6 +2590,7 @@ static int spi_nor_select_erase(struct spi_nor *nor)
 	wanted_size = 4096u;
 #endif
 
+	trace_vendor_spinor_erase_size(nor, &wanted_size);
 	if (spi_nor_has_uniform_erase(nor)) {
 		erase = spi_nor_select_uniform_erase(map, wanted_size);
 		if (!erase)
diff --git a/include/trace/hooks/spinor.h b/include/trace/hooks/spinor.h
new file mode 100644
index 000000000000..68149e4adcbe
--- /dev/null
+++ b/include/trace/hooks/spinor.h
@@ -0,0 +1,46 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) Huawei Technologies Co., Ltd. 2022-2022. All rights reserved.
+ * Description: Spinor driver Vendor Hooks
+ */
+
+#ifdef CONFIG_VENDOR_SPINOR_HOOKS
+#undef TRACE_SYSTEM
+#define TRACE_SYSTEM spinor
+
+#define TRACE_INCLUDE_PATH trace/hooks
+#if !defined(_TRACE_HOOK_SPINOR_HOOKS_H) || defined(TRACE_HEADER_MULTI_READ)
+#define _TRACE_HOOK_SPINOR_HOOKS_H
+#include <linux/tracepoint.h>
+#include <trace/hooks/vendor_hooks.h>
+
+struct spi_nor;
+struct flash_info;
+
+DECLARE_HOOK(vendor_spinor_erase_size,
+	TP_PROTO(struct spi_nor *nor, unsigned int *size),
+	TP_ARGS(nor, size));
+
+DECLARE_HOOK(vendor_spi_nor_read_id,
+	TP_PROTO(const struct flash_info **info),
+	TP_ARGS(info));
+
+#endif
+/* This part must be outside protection */
+
+#include <trace/define_trace.h>
+
+#else
+
+#include <linux/tracepoint.h>
+struct spi_nor;
+struct flash_info;
+DECLARE_EVENT_NOP(vendor_spinor_erase_size,
+	TP_PROTO(struct spi_nor *nor, unsigned int *size),
+	TP_ARGS(nor, size));
+
+DECLARE_EVENT_NOP(vendor_spi_nor_read_id,
+	TP_PROTO(const struct flash_info **info),
+	TP_ARGS(info));
+
+#endif
-- 
2.17.1

